libnetwork/resolver.go

```diff
@@ -479,19 +479,6 @@ func (r *Resolver) forwardExtDNS(proto string, query *dns.Msg) *dns.Msg {
 			// Server returned REFUSED: this can be a transitional status, so continue with the next external DNS server
 			logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), queryName)
 			continue
-		case dns.RcodeNameError:
-			// Server returned NXDOMAIN. Stop resolution if it's an authoritative answer (see RFC 8020: https://tools.ietf.org/html/rfc8020#section-2)
-			logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), queryName)
-			if resp.Authoritative {
-				break
-			}
-			continue
-		case dns.RcodeSuccess:
-			// All is well
-		default:
-			// Server gave some error. Log the error, and continue with the next external DNS server
-			logrus.Debugf("[resolver] external DNS %s:%s responded with %s (code %d) for %q", proto, extDNS.IPStr, statusString(resp.Rcode), resp.Rcode, queryName)
-			continue
```

Comment:
> Should we still log this?

I see no value in logging responses which are forwarded verbatim to the client.

> Do we know what other errors we could receive?

[Yes.](https://pkg.go.dev/github.com/miekg/dns#RcodeSuccess)

> Would it still make sense to fall back to the next DNS, in case the first had an issue?

No, because the only response codes indicating a server issue are already handled by falling back to the next server. Doing the same for any other response code would be like an HTTP reverse proxy falling back on the next server because the first replied with a 4xx response.

Reply:
> I see no value in logging responses which are forwarded verbatim to the client.

These logs for sure have helped debugging in the past; they get handled inside the container by `<whatever>` runs inside the container, and having logs on the daemon side could help debugging "hey, it's an issue with your DNS"

Reply:
>> Do we know what other errors we could receive?
>
> [Yes.](https://pkg.go.dev/github.com/miekg/dns#RcodeSuccess)

>> Would it still make sense to fall back to the next DNS, in case the first had an issue?
>
> No, because the only response codes indicating a server issue are already handled by falling back to the next server. Doing the same for any other response code would be like an HTTP reverse proxy falling back on the next server because the first replied with a 4xx response.

I guess my question here is are those 4xx responses equivalent for all configured DNS servers? e.g. looking at `VERSION`, `BADALG`, `BADTIME`, and without fully digging through them, some of those (at a glance) seem like "this DNS does not implement X" (but the next DNS that's configured _may_ ?)

Reply:
> Unless we dump the entire query and response in the logs, comparing dig vs dig @upstream.dns.server from inside the container would be much more informative for troubleshooting DNS issues.

Reality is that this will be a lot more involved. The usual report will be "docker's DNS doesn't resolve X".

Getting the user to provide daemon logs, and those logs showing "we tried your DNS, and your DNS gave a XX error" means we can show it's (most likely) not an issue with our embedded DNS, and that the user needs to go look elsewhere.

Reply:
> I guess my question here is are those 4xx responses equivalent for all configured DNS servers?

Yes, for the same reason as the "split horizon" misconfiguration. All configured upstream servers should be interchangeable as [failover is only for handling unresponsive servers or server failure.](https://datatracker.ietf.org/doc/html/rfc1034#autoid-41)

> The usual report will be "docker's DNS doesn't resolve X".
> 
> Getting the user to provide daemon logs, and those logs showing "we tried your DNS, and your DNS gave a XX error" means we can show it's (most likely) not an issue with our embedded DNS, and that the user needs to go look elsewhere.

And those users aren't going to be running the daemon in debug mode, so they will have to actively do things to change that and collect the logs. What's the difference if we instead instruct them to run `dig` a couple of times?

Reply:
> And those users aren't going to be running the daemon in debug mode, so they will have to actively do things to change that and collect the logs. What's the difference if we instead instruct them to run dig a couple of times?

- Minimal images without dig
- Alpine images doing different resolution than non-alpine
- "In my `<insert complex setup>`, things fail"

Whereas (and perhaps some of these logs should not be `Debug()` if they _should be_ infrequent) having a logs that proves "it's not our problem" can avoid all of that, and can save us a lot of time. Daemon logs are also collected in Docker Desktop support dumps, so can provide some of this information after-the-fact.

Reply:
https://github.com/moby/moby/blob/de9f12ea962608bd257a07af72a792eba2a38132/libnetwork/resolver.go#L512-L514
We already have a debug log for every reply with no answer section. Is that not sufficient?